File: streamflow-core/streamflow-model/src/main/java/streamflow/model/Resource.java
/**
 * Copyright 2014 Lockheed Martin Corporation
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package streamflow.model;

import com.fasterxml.jackson.databind.annotation.JsonDeserialize;
import com.fasterxml.jackson.databind.annotation.JsonSerialize;
import org.mongodb.morphia.annotations.Embedded;
import org.mongodb.morphia.annotations.Entity;
import org.mongodb.morphia.annotations.Id;
import streamflow.model.util.DateSerializer;
import java.io.Serializable;
import java.util.Date;
import streamflow.model.util.DateDeserializer;

@Entity("resource")
public class Resource implements streamflow.model.util.Entity<String>, Serializable {

    @Id
    private String id;

    private String name;

    private String label;

    private Date created;

    private Date modified;

    @Embedded
    private ResourceConfig config;

    private String framework;
    
    private String frameworkLabel;

    private String version;

    
    public Resource() {
    }

    @Override
    public String getId() {
        return id;
    }

    public void setId(String id) {
        this.id = id;
    }

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }

    public String getLabel() {
        return label;
    }

    public void setLabel(String label) {
        this.label = label;
    }

    public String getFramework() {
        return framework;
    }

    public void setFramework(String framework) {
        this.framework = framework;
    }

    public String getFrameworkLabel() {
        return frameworkLabel;
    }

    public void setFrameworkLabel(String frameworkLabel) {
        this.frameworkLabel = frameworkLabel;
    }

    public String getVersion() {
        return version;
    }

    public void setVersion(String version) {
        this.version = version;
    }

    @JsonSerialize(using = DateSerializer.class)
    public Date getCreated() {
        return created;
    }

    @JsonDeserialize(using = DateDeserializer.class)
    public void setCreated(Date created) {
        this.created = created;
    }

    @JsonSerialize(using = DateSerializer.class)
    public Date getModified() {
        return modified;
    }

    @JsonDeserialize(using = DateDeserializer.class)
    public void setModified(Date modified) {
        this.modified = modified;
    }

    public ResourceConfig getConfig() {
        return config;
    }

    public void setConfig(ResourceConfig config) {
        this.config = config;
    }

    @Override
    public int hashCode() {
        int hash = 3;
        hash = 89 * hash + (this.id != null ? this.id.hashCode() : 0);
        hash = 89 * hash + (this.name != null ? this.name.hashCode() : 0);
        hash = 89 * hash + (this.label != null ? this.label.hashCode() : 0);
        hash = 89 * hash + (this.created != null ? this.created.hashCode() : 0);
        hash = 89 * hash + (this.modified != null ? this.modified.hashCode() : 0);
        hash = 89 * hash + (this.config != null ? this.config.hashCode() : 0);
        hash = 89 * hash + (this.framework != null ? this.framework.hashCode() : 0);
        hash = 89 * hash + (this.frameworkLabel != null ? this.frameworkLabel.hashCode() : 0);
        hash = 89 * hash + (this.version != null ? this.version.hashCode() : 0);
        return hash;
    }

    @Override
    public boolean equals(Object obj) {
        if (obj == null) {
            return false;
        }
        if (getClass() != obj.getClass()) {
            return false;
        }
        final Resource other = (Resource) obj;
        if ((this.id == null) ? (other.id != null) 
                : !this.id.equals(other.id)) {
            return false;
        }
        if ((this.name == null) ? (other.name != null) 
                : !this.name.equals(other.name)) {
            return false;
        }
        if ((this.label == null) ? (other.label != null) 
                : !this.label.equals(other.label)) {
            return false;
        }
        if (this.created != other.created 
                && (this.created == null || !this.created.equals(other.created))) {
            return false;
        }
        if (this.modified != other.modified 
                && (this.modified == null || !this.modified.equals(other.modified))) {
            return false;
        }
        if (this.config != other.config 
                && (this.config == null || !this.config.equals(other.config))) {
            return false;
        }
        if ((this.framework == null) ? (other.framework != null) 
                : !this.framework.equals(other.framework)) {
            return false;
        }
        if ((this.frameworkLabel == null) ? (other.frameworkLabel != null) 
                : !this.frameworkLabel.equals(other.frameworkLabel)) {
            return false;
        }
        if ((this.version == null) ? (other.version != null) 
                : !this.version.equals(other.version)) {
            return false;
        }
        return true;
    }

    @Override
    public String toString() {
        return "Resource{" + "id=" + id + ", name=" + name + ", label=" + label 
                + ", created=" + created + ", modified=" + modified + ", config=" + config 
                + ", framework=" + framework + ", frameworkLabel=" + frameworkLabel 
                + ", version=" + version + '}';
    }
}

File: streamflow-core/streamflow-server/src/main/webapp/app/resource/resource.js
/*
 * Copyright 2014 Lockheed Martin Corporation
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
var resourceModule = angular.module('streamflow.resource',
        ['ngResource', 'ui.bootstrap', 'streamflow.notify']);

// Route configuration for the various views
resourceModule.config(['$routeProvider', function($routeProvider) {
    // Set up routes for the specific paths to the different views
    $routeProvider.
            when('/resources', {
                templateUrl: 'app/resource/resource.tpl.list.html',
                controller: 'ResourceListController'
            }).
            when('/resources/:id', {
                templateUrl: 'app/resource/resource.tpl.view.html',
                controller: 'ResourceViewController'
            }).
            when('/resources/:id/entries/:entryId', {
                templateUrl: 'app/resource/resource.tpl.entry.view.html',
                controller: 'ResourceEntryViewController'
            });
}]);


////////////////////////////////////////////////////////////////////////////////
// SERVICES
////////////////////////////////////////////////////////////////////////////////

resourceModule.factory('Resource', ['$resource', function($resource) {
    return $resource('api/resources/:id', {id: '@id'}, {
        getConfig: {
            method: 'GET',
            url: 'api/resources/:id/config'
        }
    });
}]);

resourceModule.factory('ResourceEntry', ['$resource', function($resource) {
    return $resource('api/resources/:id/entries/:entryId', {id: '@id', entryId: '@entryId'}, {
        update: {
            method: 'PUT'
        },
        remove: {
            method: 'DELETE'
        },
        getConfig: {
            method: 'GET',
            url: 'api/resources/:id/entries/:entryId/config'
        },
        updateConfig: {
            method: 'PUT',
            url: 'api/resources/:id/entries/:entryId/config'
        }
    });
}]);


////////////////////////////////////////////////////////////////////////////////
// CONTROLLERS
////////////////////////////////////////////////////////////////////////////////

resourceModule.controller('ResourceListController', [
    '$scope', '$modal', 'Resource',
    function($scope, $modal, Resource) {

        $scope.listResources = function() {
            $scope.resources = Resource.query(
                function() {
                    $scope.resourcesMessage = 'No resources were found';
                },
                function() {
                    $scope.resourcesMessage = 'Unable to load the resource list';
                }
            );
        };

        $scope.newResourceEntry = function(resource) {
            // Open the submit topology dialog to deploy the selected topology
            $modal.open({
                templateUrl: 'app/resource/resource.tpl.entry.create.html',
                controller: 'ResourceEntryCreateController',
                resolve: {
                    resource: function() {
                        return angular.copy(resource);
                    }
                }
            });
        };

        $scope.listResources();
    }
]);

resourceModule.controller('ResourceViewController', [
    '$scope', '$routeParams', '$modal', 'streamflowNotify', 'Resource', 'ResourceEntry',
    function($scope, $routeParams, $modal, streamflowNotify, Resource, ResourceEntry) {
        $scope.resource = Resource.get({id: $routeParams.id});

        $scope.listResourceEntries = function() {
            $scope.resourceEntries = ResourceEntry.query({id: $routeParams.id},
                function() {
                    $scope.resourceEntriesMessage = 'No resource entries were found';
                },
                function() {
                    $scope.resourceEntriesMessage = 'Unable to load the resource entry list';
                }
            );
        };

        $scope.newResourceEntry = function() {
            $modal.open({
                templateUrl: 'app/resource/resource.tpl.entry.create.html',
                controller: 'ResourceEntryCreateController',
                resolve: {
                    resource: function() {
                        return $scope.resource;
                    }
                }
            }).result.then(function() {
                $scope.listResourceEntries();
            });
        };

        $scope.deleteResourceEntry = function(resourceEntry) {
            $modal.open({
                templateUrl: 'app/resource/resource.tpl.entry.delete.html',
                controller: function($scope) {
                    $scope.resourceEntry = resourceEntry;
                }
            }).result.then(function() {
                // Delete the resource entry on the server
                ResourceEntry.remove({id: resourceEntry.resource, entryId: resourceEntry.id},
                    function() {
                        streamflowNotify.success('The resource entry was deleted successfully');
                    },
                    function() {
                        streamflowNotify.error('The resurce entry was not deleted due to a server error.');
                    }
                );

                // After deleting the resource entry, refresh the entries list 
                $scope.listResourceEntries();
            });
        };

        $scope.listResourceEntries();
    }
]);

resourceModule.controller('ResourceEntryViewController',  [
    '$scope', '$location', '$routeParams', '$modal', 'streamflowNotify', 'Resource', 'ResourceEntry',
    function($scope, $location, $routeParams, $modal, streamflowNotify, Resource, ResourceEntry) {
        $scope.resource = Resource.get({id: $routeParams.id});
        $scope.resourceEntry = ResourceEntry.get({id: $routeParams.id, entryId: $routeParams.entryId});

        // Load the user saved resource entry config object
        $scope.resourceEntryConfig = ResourceEntry.getConfig({
            id: $routeParams.id, entryId: $routeParams.entryId},
        function() {
            // Make sure the resource entry config was initialized properly
            if (!$scope.resourceEntryConfig) {
                $scope.resourceEntryConfig = {};
            }
            if (!$scope.resourceEntryConfig.properties) {
                $scope.resourceEntryConfig.properties = {};
            }

            // Retrieve the resource config after the user defined config is loaded
            $scope.resourceConfig = Resource.getConfig({id: $routeParams.id},
            function(resourceConfig) {
                // Iterate over all of the properties to ensure the resource entry config is synced
                angular.forEach(resourceConfig.properties, function(resourceProperty) {
                    // Check if the user config already has a property from the resource config
                    if (!$scope.resourceEntryConfig.properties[resourceProperty.name]) {
                        if (resourceProperty.defaultValue) {
                            $scope.resourceEntryConfig.properties[resourceProperty.name] =
                                    resourceProperty.defaultValue;
                        } else {
                            $scope.resourceEntryConfig.properties[resourceProperty.name] = '';
                        }
                    }
                });
            }
            );
        }
        );

        $scope.saveResourceEntry = function() {
            // Post the update topology config to the server
            ResourceEntry.updateConfig({id: $routeParams.id, entryId: $routeParams.entryId}, $scope.resourceEntryConfig,
                function() {
                    streamflowNotify.success('The resource entry was saved successfully.');

                    // Redirect back to the topology view if saved successfully
                    $location.path('/resources/' + $routeParams.id);
                },
                function() {
                    streamflowNotify.error('The resource entry was not saved due to a server error..');
                }
            );
        };

        $scope.cancelResourceEntry = function() {
            // Redirect back to the resources view without saving if cancelled
            $location.path('/resources/' + $routeParams.id);
        };
    }
]);

/**
 * Topology controller used in the topology-create dialog to create a new topology
 */
resourceModule.controller('ResourceEntryCreateController', [
    '$scope', '$location', '$modalInstance', 'streamflowNotify', 'resource', 'ResourceEntry',
    function($scope, $location, $modalInstance, streamflowNotify, resource, ResourceEntry) {
        
        $scope.create = function() {
            console.log("Resource ID = " + resource.id);
            console.log('Name = ' + $scope.resourceEntryName);
            
            // Create the new topology using the specified paramters from the dialog
            var resourceEntry = new ResourceEntry();
            resourceEntry.name = $scope.resourceEntryName;
            resourceEntry.description = $scope.resourceEntryDescription;
            resourceEntry.resource = resource.id;
            resourceEntry.config = {};

            resourceEntry.$save({id: resource.id},
                function(resourceEntry) {
                    // After the topology is created, redirect the user to the new empty topology view
                    $location.path('/resources/' + resource.id + '/entries/' + resourceEntry.id);

                    // Close the dialog after successful creation
                    $modalInstance.close();
                },
                function() {
                    streamflowNotify.error('Unable to create resource entry due to server error: ', 'error');
                }
            );
        };

        $scope.cancel = function() {
            $modalInstance.dismiss();
        };
    }
]);


////////////////////////////////////////////////////////////////////////////////
// DIRECTIVES
////////////////////////////////////////////////////////////////////////////////
